### PR TITLE
[python-package] Load parameters from model string

### LINKED ISSUE #6851: [python-package] Model parameters are lost when loading Booster from model serialized string
## Description

After training, saving the model on a string, then retrieving it, the `params` dictionary are lost

## Reproducible example

```python
import lightgbm as lgb
import numpy as np

# Train data
train_x = np.random.rand(1000, 20)
train_y = np.random.randint(0, 1, 1000)
train_data = lgb.Dataset(train_x, train_y)

# Parameters
params = {
    "boosting_type": "gbdt",
    "objective": "binary",
    "metric": "auc",
    "num_leaves": 31,
    "learning_rate": 0.05,
    "feature_fraction": 0.9,
    "bagging_fraction": 0.8,
    "bagging_freq": 5,
    "verbose": -100,
}

# train the model
model = lgb.train(params, train_data)

# print the params, should be the same as above
print(model.params)

# dump the model into a string
model_serialized = model.model_to_string()

# load a new model with the string
new_model = lgb.Booster(model_str=model_serialized)

# print the params, it returns an empty dict
print(new_model.params)

```

## Environment info

LightGBM version or commit hash: 4.6.0

Command(s) you used to install LightGBM

```shell
$ pip install lightgbm
```

ArchLinux system

## Additional Comments

I think the fix is pretty simple and stems from the `__init__()` function last line where the `params` member is [overwritten with whatever is passed (empty, when loading from a serialized string)](https://github.com/microsoft/LightGBM/blob/f7c641d980a1a40df5395cd9b12589834860879a/python-package/lightgbm/basic.py#L3722)

I guess the fix is easy, will PR it in a moment.

### PR DESCRIPTION
**Pull request by Samsagax**
This fixes #6851 by using the same workaround as when loading the model from a file (#5424).

### CODE REVIEW

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks for taking the time to contribute!

As I just mentioned in https://github.com/microsoft/LightGBM/issues/6851#issuecomment-2692937425, think the issue goes deeper than just this and that #6101 will be the right fix.

That said, if you want to pursue this PR for just the in-memory string part (not also fixing `copy()`), then I'm open to it.... but please add a new test covering this behavior.

Otherwise, if you don't have the time/interest in that, we can close this and you can subscribe to #6101.

**Comment by Samsagax** [comment]
@microsoft-github-policy-service agree

**Comment by Samsagax** [comment]
Ok. Test case added. Should also be useful for #6101 

Thanks for having the time to review this.

File: tests/python_package_test/test_basic.py [hunk 1]
```diff
@@ -997,3 +997,39 @@ def test_equal_datasets_from_one_and_several_matrices_w_different_layouts(rng, t
...
+    model = lgb.train(params, train_data)
+    model_serialized = model.model_to_string()
+
+    # load a new model with the string
+    new_model = lgb.Booster(model_str=model_serialized)
+
+    assert(new_model.params) # Check not empty dict
+    # check the params we provided
+    print(new_model.params)
+    print(params)
```
Comment (jameslamb) [inline tests/python_package_test/test_basic.py:1031]:
```suggestion
```

This looks like it's left over from debugging. Please remove it.

File: tests/python_package_test/test_basic.py [hunk 2]
```diff
@@ -997,3 +997,39 @@ def test_equal_datasets_from_one_and_several_matrices_w_different_layouts(rng, t
...
+
+def test_string_serialized_params_retrieval(rng):
+    # Random train data
+    train_x = rng.random((500, 20))
+    train_y = rng.integers(0, 1, 500)
+    train_data = lgb.Dataset(train_x, train_y)
+
+    # Parameters
+    params = {
+        "boosting": "gbdt",
```
Comment (jameslamb) [inline tests/python_package_test/test_basic.py:1010]:
```suggestion
        "boosting": "gbdt",
        "deterministic": true,
        "feature_contri": [0.5] * train_x.shape[1],
        "interaction_constraints": [[0, 1], [0]],
```

This test is currently checking that the following types of parameters successfully make it through this serialization-deserialization round trip:

* string
* int
* float
* array of string

Let's please also test other types:

* boolean
* array of float
* array of arrays of integers

File: tests/python_package_test/test_basic.py [hunk 3]
```diff
@@ -997,3 +997,39 @@ def test_equal_datasets_from_one_and_several_matrices_w_different_layouts(rng, t
...
+        "num_leaves": 31,
+        "learning_rate": 0.05,
+        "feature_fraction": 0.9,
+        "bagging_fraction": 0.8,
+        "bagging_freq": 5,
+        "verbosity": -100,
+    }
+
+    # train a model and serialize it to a string in memory
+    model = lgb.train(params, train_data)
```
Comment (jameslamb) [inline tests/python_package_test/test_basic.py:1022]:
```suggestion
        "num_leaves": 7
        "learning_rate": 0.05,
        "feature_fraction": 0.9,
        "bagging_fraction": 0.8,
        "bagging_freq": 5,
        "verbosity": -100,
    }

    # train a model and serialize it to a string in memory
    model = lgb.train(params, train_data, num_boost_round=2)
```

Let's please train an even smaller, cheaper model. The model's statistical performance and total number of leaves isn't really important here, since this is about how parameters survive a serialization-deserialization loop.

File: tests/python_package_test/test_basic.py [hunk 4]
```diff
@@ -997,3 +997,39 @@ def test_equal_datasets_from_one_and_several_matrices_w_different_layouts(rng, t
...
+    # load a new model with the string
+    new_model = lgb.Booster(model_str=model_serialized)
+
+    assert(new_model.params) # Check not empty dict
+    # check the params we provided
+    print(new_model.params)
+    print(params)
+    for k, v in params.items():
+        assert(k in new_model.params)
+        assert(new_model.params[k] == v)
```
Comment (jameslamb) [inline tests/python_package_test/test_basic.py:1034]:
```suggestion
    assert new_model["objective"] == "binary"
    assert new_model["metric"] == "auc"
```

Can you please replace this `for` loop with explicit `assert` statements like this, one per parameter? In library code it's valuable to reduce duplication as you have, but here in test code I'd prefer the strictness and safety we get from duplication.

`params` is a Python dictionary, which is mutable. Repeating the expected values with literals here reduces the risk of this test silently and incorrectly passing if a bug in `lightgbm` resulted in `params` being modified in-place by `lgb.train()`.

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks very much! I'm glad this worked and I think it's a really nice improvement.

I left some suggestions to make the tests stricter. Can you please also move that test case to right after the test on the similar logic for loading from a model file?

https://github.com/microsoft/LightGBM/blob/6437645c4a0c17046be59e4f57d09952e2e0185f/tests/python_package_test/test_engine.py#L1461

File: tests/python_package_test/test_basic.py (hunk 4 above)
Comment (Samsagax) [reply tests/python_package_test/test_basic.py]: Done

File: tests/python_package_test/test_basic.py [hunk 5]
```diff
@@ -997,3 +997,39 @@ def test_equal_datasets_from_one_and_several_matrices_w_different_layouts(rng, t
...
+
```
[the remark this answers is not in the document]
Comment (Samsagax) [reply tests/python_package_test/test_basic.py]: Done

File: tests/python_package_test/test_basic.py [hunk 6]
```diff
@@ -997,3 +997,4 @@ def test_equal_datasets_from_one_and_several_matrices_w_different_layouts(rng, t
     lgb.Dataset(mat)._dump_text(one_path)
 
     assert filecmp.cmp(one_path, several_path)
+
```
Comment (jameslamb) [inline tests/python_package_test/test_basic.py:1000]:
```suggestion
```

Please remove this empty line.

File: tests/python_package_test/test_engine.py [hunk 7]
```diff
@@ -1508,6 +1509,52 @@ def test_parameters_are_loaded_from_model_file(tmp_path, capsys, rng):
...
+    model_serialized = model.model_to_string()
+
+    # load a new model with the string
+    with pytest.warns(UserWarning,
+                      match="Ignoring params argument, using parameters from model string."):
+        new_model = lgb.Booster(params={"num_leaves": 7}, model_str=model_serialized)
+
+    assert(new_model.params) # Check not empty dict
+    assert(new_model.params["boosting"] == "gbdt")
+    assert(new_model.params["deterministic"] == True)
```
Comment (jameslamb) [inline tests/python_package_test/test_engine.py]:
```suggestion
    assert(new_model.params["deterministic"] is True)
```

This should resolve the linter error:

> tests/python_package_test/test_engine.py:1545:12: E712 Avoid equality comparisons to `True`; use `if new_model.params["deterministic"]:` for truth checks

([build link](https://github.com/microsoft/LightGBM/actions/runs/13703032288/job/38321483346?pr=6852#step:3:898))

File: tests/python_package_test/test_engine.py [hunk 8]
```diff
@@ -1508,6 +1509,52 @@ def test_parameters_are_loaded_from_model_file(tmp_path, capsys, rng):
...
+    # train a model and serialize it to a string in memory
+    model = lgb.train(params, train_data, num_boost_round=2)
+    model_serialized = model.model_to_string()
+
+    # load a new model with the string
+    with pytest.warns(UserWarning,
+                      match="Ignoring params argument, using parameters from model string."):
+        new_model = lgb.Booster(params={"num_leaves": 7}, model_str=model_serialized)
+
+    assert(new_model.params) # Check not empty dict
```
Comment (jameslamb) [inline tests/python_package_test/test_engine.py:1543]:
```suggestion
```

I think this check is not necessary, as all of the others below could only be true if the dictionary was non-empty. Let's remove it to simplify this test a bit.

**Review by jameslamb — DISMISSED** [review]
Thanks, looks great!

As a quick note... I already see the linting failed: https://github.com/microsoft/LightGBM/actions/runs/13703032288/job/38321483346?pr=6852

You can replicate that locally (and auto-fix the formatting issues) like this:

```shell
pip install pre-commit
pre-commit run --all-files
```

No need to squash commits... we squash on merges here, so every pull request = 1 commit on `master`: https://github.com/microsoft/LightGBM/commits/master/

You should merge in the latest changes whenever updating though... that helps ensure this is tested against the latest state of the target branch. I just did that, by clicking this button:

image

File: tests/python_package_test/test_engine.py [hunk 9]
```diff
@@ -1508,6 +1509,52 @@ def test_parameters_are_loaded_from_model_file(tmp_path, capsys, rng):
...
+    assert(new_model.params["boosting"] == "gbdt")
+    assert(new_model.params["deterministic"])  # Linter doesn't like boolean compare
```
Comment (jameslamb) [inline tests/python_package_test/test_engine.py:1545]:
```suggestion
    assert(new_model.params["deterministic"] is True)
```

Ah you'd addressed this before I published my review. Let's see if a comparison like `is True` will make it happy.

If not, then leave it as `is True` and add a comment like this:

```python
assert(new_model.params["deterministic"])  # noqa: E712
```

**Comment by Samsagax** [comment]
> As a quick note... I already see the linting failed: https://github.com/microsoft/LightGBM/actions/runs/13703032288/job/38321483346?pr=6852
> 
> You can replicate that locally (and auto-fix the formatting issues) like this:
> 
> ```shell
> pip install pre-commit
> pre-commit run --all-files
> ```

Thanks

> No need to squash commits... we squash on merges here, so every pull request = 1 commit on `master`: https://github.com/microsoft/LightGBM/commits/master/
> 
> You should merge in the latest changes whenever updating though... that helps ensure this is tested against the latest state of the target branch. I just did that, by clicking this button:

I prefer rebasing and rewriting it, but each project has it's own nit-picks. Will do.

Thanks again. The latest version will come in shortly

**Comment by Samsagax** [comment]
Don't really know why r packaging is failing. But seems to be failing in master too.

**Comment by jameslamb** [comment]
> Don't really know why r packaging is failing. But seems to be failing in master too.

Yes you're right, that job failing is unrelated to your changes. Documented at #6855

**Review by jameslamb — COMMENTED** [review]
Thanks very much! I've changed the label to `breaking`, because this is technically a breaking change from the perspective of anyone who was relying on `params` passed through the `Booster` constructor when loading a file, like this:

```python
bst = Booster(model_str=model_str, params={...})
```

@jmoralez do you have some time to review this too? You're so familiar with these code paths from #5424, I think that'd be really helpful.

I approve of these changes as-is (so nothing else you need to do @Samsagax), but just leaving a "comment" review so my approval won't count towards a merge until we get another approval.

File: tests/python_package_test/test_engine.py [hunk 10]
```diff
@@ -1508,6 +1508,50 @@ def test_parameters_are_loaded_from_model_file(tmp_path, capsys, rng):
...
+        "verbosity": -100,
+    }
+
+    # train a model and serialize it to a string in memory
+    model = lgb.train(params, train_data, num_boost_round=2)
+    model_serialized = model.model_to_string()
+
+    # load a new model with the string
+    with pytest.warns(UserWarning, match="Ignoring params argument, using parameters from model string."):
+        new_model = lgb.Booster(params={"num_leaves": 7}, model_str=model_serialized)
```
Comment (jmoralez) [inline tests/python_package_test/test_engine.py:1539]: I think it'd be helpful to set a different value here to make sure that it's actually ignored, e.g. `"num_leaves": 35`

Comment (Samsagax) [reply tests/python_package_test/test_engine.py]: Done

**Comment by Samsagax** [comment]
Don't know why is failing. But seems the same failure as main.

**Review by jmoralez — APPROVED** [review]

**Comment by jameslamb** [comment]
> But seems the same failure as main.

Yes, you don't need to do anything else at this point. There are a couple of CI issues here, we're working on them in #6872

**Review by StrikerRUS — APPROVED** [review]
LGTM, thanks!
